raft/server.cc

```diff
@@ -1267,20 +1259,30 @@ future<> server_impl::enter_joint_configuration(server_address_set c_new, seasta
     if (joining.size() == 0 && leaving.size() == 0) {
         co_return;
     }
+
     _stats.add_config++;
-    co_await add_entry_internal(raft::configuration{std::move(c_new)}, wait_type::committed, as);
-}
+    const auto& e = _fsm->add_entry(raft::configuration{std::move(c_new)});
 
-future<> server_impl::set_configuration(server_address_set c_new, seastar::abort_source* as) {
-    co_await enter_joint_configuration(std::move(c_new), as);
-    // Above we co_wait that the joint configuration is committed.
+    // We've just submitted a joint configuration to be committed.
     // Immediately, without yield, once the FSM discovers
-    // this, it appends non-joint entry. Hence,
-    // at this point in execution, non-joint entry is in the log.
-    // By waiting for a follow up dummy to get committed, we
-    // automatically wait for the non-joint entry to get
-    // committed.
-    co_await add_entry_internal(log_entry::dummy(), wait_type::committed, as);
+    // joint configuration, it appends a corresponding non-joint entry.
+    // By waiting for the joint configuration first we ensure
+    // that the next non-joint configuration we get from fsm in io_fiber
+    // would be the one corresponding to our joint configuration,
+    // no matter if the leader changed in the meantime.
```

Comment:
The comment above is incorrect after the change. Before the change the reason the comment was correct is because we co_awaited for the joint entry to be committed (as first deleted line says), but at the point the comment is now we did not co_await for anything and it is definitely incorrect to say that after adding a joint entry to be committed "Immediately, without yield, once the FSM discovers joint configuration, it appends a corresponding non-joint entry." In fact there may be a lot of yields in between.

Reply:
> This comment is for the next block of code.

The comment states what fsm does at the point of the comment. What it states is wrong at this point.

> 
> Could you elaborate, why the statement
> 
> > after adding a joint entry to be committed "Immediately, without yield, once the FSM discovers joint configuration, it appends a corresponding non-joint entry."
> 
> is false? This logic hasn't changed with this commit.

Overall logic did not. The placement of the comment did. You replaced  "Above we co_wait that the joint configuration is committed." with "We've just submitted a joint configuration to be committed." and this is two very different things. You cannot expect that the sentence after that "Immediately, without yield, once the FSM discovers" holds. And indeed it is not. What comment states is correct after co_await, not before.

Reply:
I guess the problem is with the words "once the FSM discovers joint configuration", they refer to the moment when the joint configuration is committed, not when it's submitted. Should be changed to "once the FSM discovers a committed joint configuration".

Reply:
I would also drop "Immediately, without yield". Do not see how it is relevant now.